### PR TITLE
Merge outstanding commits to `dev`

### PR DESCRIPTION
**Pull request by tomlogic**
This first PR is to get outstanding commits merged into the `dev` branch before I start work to remove the `yaml-cpp` dependency for `pinproctest`.  I may also make additional `cmake` changes to ensure it builds properly on Linux and Windows with either MinGW/gcc (under both bash and a command prompt?) or Visual Studio.

Although I merged work from @Compy and @jabdoa2, I did revert the commits that added Debian control files and that disabled the `pinproctest` example program.  Debian is (I believe) incomplete and includes generated files that shouldn't be in the repository, and it's already possible to skip the examples by setting `-D PINPROC_BUILD_TOOLS=OFF` in the `cmake` configuration.

I left out two commits by @koenheltzel because I believe they're covered by other commits.  The addition of `C:` to the paths is handled by the `CROSS_ROOT` option.  I guess we could default that to `C:` for `WIN32` if desired.

And I left out a commit by @bingopodcast because part of it is covered by setting `EXTRA_INC` and I was unsure of the necessity of `set(lib_ftdi_usb usb-1.0 ftdi1)` for all builds.

And I'm tagging @darren5 and @gstellenberg because I included their commits.  I Gerry was concerned that his change to `pinprocfw` wouldn't work for P-ROC, but I successfully tested it.

I welcome all feedback.  Keep in mind that after merging this I'm planning on another PR to remove yaml-cpp entirely, so please only report build problems that exist with `PINPROC_BUILD_TOOLS=OFF`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -13,12 +13,18 @@ if(POLICY CMP0015)
 	cmake_policy(SET CMP0015 OLD)
 endif()
 
+# allow relative paths in LINK_DIRECTORIES
+if(POLICY CMP0081)
+	cmake_policy(SET CMP0081 OLD)
+endif()
```
Comment (tomlogic) [inline CMakeLists.txt:19]: This addition made it possible to specify relative `EXTRA_LINK` directories in addition to `EXTRA_INC`.

File: CMakeLists.txt [hunk 2]
```diff
@@ -13,12 +13,18 @@ if(POLICY CMP0015)
...
 
 ###
 ### Project settings
 ###
 project(PINPROC)
 
+set (CMAKE_CXX_STANDARD 11)
+
```
Comment (tomlogic) [inline CMakeLists.txt:27]: This addition was for updated `yaml-cpp`.  I may revert the change after removing the `yaml-cpp` dependency.

File: CMakeLists.txt [hunk 3]
```diff
@@ -69,8 +76,15 @@ if(VERBOSE)
 endif()
 
 # use -DEXTRA_INC="<path>;<path>" and -DEXTRA_LINK="<path>;<path>"
+set(EXTRA_INC "" CACHE STRING "Extra include directories separated by ;")
+set(EXTRA_LINK "" CACHE STRING "Extra link directories separated by ;")
```
Comment (tomlogic) [inline CMakeLists.txt:80]: I wasn't able to pass these options (`EXTRA_INC` and `EXTRA_LINK`) on the command line until adding them here.  I'm not a CMake expert so feedback on ways to improve this are welcome.

File: CMakeLists.txt [hunk 4]
```diff
@@ -69,8 +76,15 @@ if(VERBOSE)
...
+if(CROSS_ROOT)
+include_directories(${PINPROC_SOURCE_DIR}/include ${EXTRA_INC} ${CROSS_ROOT}/usr/local/include)
+link_directories(${EXTRA_LINK} ${CROSS_ROOT}/usr/local/lib)
+else()
```
Comment (tomlogic) [inline CMakeLists.txt:84]: You can set `CROSS_ROOT` to `"C:"` if necessary for building under Windows.

File: examples/pinproctest/pinproctest.cpp [hunk 5]
```diff
@@ -46,12 +46,8 @@ PRResult LoadConfiguration(YAML::Node& yamlDoc, const char *yamlFilePath)
...
             return kPRFailure;
         }
-        YAML::Parser parser(fin);
         
-        while(parser) 
-        {
-            parser.GetNextDocument(yamlDoc);
-        }
+        yamlDoc = YAML::Load(fin);
     }
```
Comment (tomlogic) [inline examples/pinproctest/pinproctest.cpp:51]: The changes in this file were related to updates in `yaml-cpp`, but will go away when I remove that feature from `pinproctest`.

File: include/pinproc.h [hunk 6]
```diff
@@ -138,6 +138,9 @@ PINPROC_API PRResult PRFlushWriteData(PRHandle handle);
 /** Write data out to the P-ROC immediately (does not require a call to PRFlushWriteData). */
 PINPROC_API PRResult PRWriteData(PRHandle handle, uint32_t moduleSelect, uint32_t startingAddr, int32_t numWriteWords, uint32_t * writeBuffer);
 
+/** Write data buffered to P-ROC (does require a call to PRFlushWriteData). */
+PINPROC_API PRResult PRWriteDataUnbuffered(PRHandle handle, uint32_t moduleSelect, uint32_t startingAddr, int32_t numWriteWords, uint32_t * writeBuffer);
+
```
Comment (tomlogic) [inline include/pinproc.h:143]: This is a new API from @jabdoa2 and @Compy.

File: src/PRDevice.cpp [hunk 7]
```diff
@@ -851,7 +847,11 @@ PRResult PRDevice::SwitchGetStates( PREventType * switchStates, uint16_t numSwit
         }
         return kPRSuccess;
     }
-    else return kPRFailure;
+    else
+    {
+        PRSetLastErrorText("Switch response length does not match.");
+        return kPRFailure;
+    }
```
Comment (tomlogic) [inline src/PRDevice.cpp:854]: Notable change here.

File: src/PRDevice.cpp [hunk 8]
```diff
@@ -1088,11 +1088,12 @@ PRResult PRDevice::VerifyChipID()
...
             }
-            if (buffer[1] != P_ROC_CHIP_ID && buffer[1] != P3_ROC_CHIP_ID) 
+            if (buffer[1] != P_ROC_CHIP_ID && buffer[1] != P3_ROC_CHIP_ID)
             {
                 DEBUG(PRLog(kPRLogError, "Error in VerifyID(): Dumping buffer\n"));
-                for (i = 0; i < bufferWords; i++) 
+                for (i = 0; i < bufferWords; i++)
                     DEBUG(PRLog(kPRLogError, "buffer[%d]: 0x%x\n", i, buffer[i]));
+                PRSetLastErrorText("Chip ID does not match.");
                 rc = kPRFailure;
```
Comment (tomlogic) [inline src/PRDevice.cpp:1097]: Some additional `PRSetLastErrorText()` calls here and below.

File: src/PRDevice.cpp [hunk 9]
```diff
@@ -1199,6 +1198,19 @@ PRResult PRDevice::WriteData(uint32_t * words, int32_t numWords)
...
+    uint32_t * buffer;
+
+    buffer = (uint32_t *)malloc((numWriteWords * 4) + 4);
+    buffer[0] = CreateBurstCommand(moduleSelect, startingAddr, numWriteWords);
+    memcpy(buffer+1, writeBuffer, numWriteWords * 4);
+    res = PrepareWriteData(buffer, numWriteWords + 1);
+    free (buffer);
+	return res;
+}
+
```
Comment (tomlogic) [inline src/PRDevice.cpp:1213]: New API, assumedly used from Python.  Nothing in `libpinproc` is making use of it.

File: src/PRDevice.cpp [hunk 10]
```diff
@@ -71,15 +71,15 @@ PRDevice* PRDevice::Create(PRMachineType machineType)
     if (machineType != kPRMachineCustom && machineType != kPRMachinePDB &&
```
Comment (tomlogic) [inline src/PRDevice.cpp:71]: The majority of changes in this file are to remove whitespace and get rid of variables used to hold return values that are then ignored.  I called out one notable change in a comment below, and there are multiple additions of `PRSetLastErrorText()` to improve error reporting.

File: src/PRHardware.cpp [hunk 11]
```diff
@@ -33,9 +33,9 @@
 #include "PRCommon.h"
```
Comment (tomlogic) [inline src/PRHardware.cpp:33]: Removes trailing whitespace and adds `PRSetLastErrorText()` calls.

File: utils/pinprocfw/pinprocfw.cpp [hunk 12]
```diff
@@ -2051,6 +2051,8 @@ int verifyP3ROCImage()
     }
 
     XSVFDBG_PRINTF( 0, "\n\nSUCCESS - Operation completed successfully.  Cycle P3-ROC power to activate any changes.\n" );
+
+    return 1;
 }
```
Comment (tomlogic) [inline utils/pinprocfw/pinprocfw.cpp:2056]: Fixes a compiler warning.  Not sure why this didn't cause problems previously.

File: utils/pinprocfw/pinprocfw.cpp [hunk 13]
```diff
@@ -2134,13 +2136,10 @@ int processFile()
...
             fclose( in );
 
-            // Destroy the P-ROC device handle:
-            PRDelete(proc);
-            proc = kPRHandleInvalid;
             return iErrorCode;
 }
 
-int checkPROCFile() {
+uint32_t checkPROCFile() {
```
Comment (tomlogic) [inline utils/pinprocfw/pinprocfw.cpp:2142]: In some of my builds I was getting a compiler warning related to the `switch` statement below.  Since we're returning a `uint32` variable in the success path, it makes sense to update the API to return that data type.

File: utils/pinprocfw/pinprocfw.cpp [hunk 14]
```diff
@@ -2176,10 +2175,20 @@ int checkPROCFile() {
...
+                    (board_rev & 0x100) >> 8;
+        fprintf(stderr, "\nReading P3-ROC board_rev: %d", board_rev);
+    }
+    else {
+        board_rev = (board_rev & 0x80) >> 7 |
+                    (board_rev & 0x40) >> 6 |
+                    (board_rev & 0x20) >> 5 |
+                    (board_rev & 0x10) >> 4;
+        fprintf(stderr, "\nReading P-ROC board_rev: %d", board_rev);
+    }
```
Comment (tomlogic) [inline utils/pinprocfw/pinprocfw.cpp:2191]: This is @gstellenberg's addition for correctly calculating the board revision for P3-ROC and P-ROC.

File: utils/pinprocfw/pinprocfw.cpp [hunk 15]
```diff
@@ -2134,13 +2136,10 @@ int processFile()
             endClock    = clock();
             fclose( in );
 
-            // Destroy the P-ROC device handle:
-            PRDelete(proc);
-            proc = kPRHandleInvalid;
             return iErrorCode;
```
Comment (tomlogic) [inline utils/pinprocfw/pinprocfw.cpp:2139]: I moved the `PRDelete()` so it's always called, even if there's an error parsing the `.p-roc` file.

File: include/pinproc.h (hunk 6 above)
Comment (jabdoa2) [reply include/pinproc.h:143]: This is used for fading in MPF (via pypinproc MPF fork with GIL fixes)

File: src/PRDevice.cpp (hunk 8 above)
Comment (jabdoa2) [reply src/PRDevice.cpp:1097]: This was to find the flasher error. I then added an error to every error path so that empty error no longer happen (hopefully).
